### PR TITLE
adding toc_height to exclude upper headings, proposal for #786 and #637

### PR DESCRIPTION
**Pull request by klml**
I added a new config value toc_height. Using an array for _toc_depth_  seems to me very unlikely. 

## test local

```
~/devel/markdown$ python3 -m unittest discover tests
... 550 tests in 1.259s
OK (skipped=71)

```

### CODE REVIEW

File: markdown/extensions/toc.py
Comment (waylan) [inline markdown/extensions/toc.py]: There must be a typo here. "from to which" can' the correct. I'm not even sure what it is supposed to say.

File: docs/extensions/toc.md
Comment (waylan) [inline docs/extensions/toc.md]: There must be a typo here. "from to which" can' the correct. I'm not even sure what it is supposed to say.

File: docs/extensions/toc.md
Comment (waylan) [inline docs/extensions/toc.md]: Sorry, but I'm struggling to understand what this is talking about. What does "will not follow" mean? I could guess, but its not clear.

**Review by waylan — CHANGES_REQUESTED** [review]
Thanks for your contribution.

 In addition to the concerns mentioned below, I'm not sure "height" is the right word to describe this feature. To me a "height" of `4` means there are four levels, but that is not what the example int he tests shows. Perhaps that is why I'm struggling to understand the documentation. I don't actually understand what you are trying to do.

Both examples in the tests simply skip level 1. Perhaps one of them should do something other than that. As  reminder, this is a feature that I find completely unnecessary, so perhaps I'm missing something obvious here. But if I am, then other users could be as well, which suggests the name and/or docs could be improved.

File: markdown/extensions/toc.py
Comment (waylan) [reply markdown/extensions/toc.py]: Also, you are getting a [flake8 error][1] here. Note that the [Contributing Guidelines][2] define how to format multi-line lists. I realize the other items in the file don't match that style, but they predate the guidelines, and as the guidelines state:

> Legacy code which does not follow the guidelines should only be updated if and when other changes (bug fix, feature addition, etc.) are being made to that section of code. 

[1]: https://travis-ci.org/Python-Markdown/markdown/jobs/490761623
[2]: https://github.com/Python-Markdown/markdown/blob/master/docs/contributing.md#code-style-guide

**Comment by klml** [comment]
@waylan thanks for your quick response.

You are right about the word height. I was unsure about the naming of ```toc_height```. I wanted to name it the [opposite of toc_depth](https://www.wordhippo.com/what-is/the-opposite-of/depth.html), and I was undecided between ```toc_height``` and ```toc_top```.
You mentioned ```skip```, but toc_depth also skips headings (at the bottom). Do you have an idea what word would be meaningful?


Sorry for all these other issues, I will get into it and get back to you with an update on my PR.

**Comment by klml** [comment]
I fixed the flake error and changed the first test to a higher toc_height.
I tried to improve the wording. But I have a little subjective arbitrability, so do you think it is now more understandable?

I am still struggling with the parameters name ```toc_height```. My alternative shortlist:


* `toc_height` is missleading, 'a "height" of `4` means there are four levels'. 
* `toc_skip` toc_depth also skips headings (at the bottom). 
* `toc_top` 
* `toc_edge` 
* `toc_begin` 
* `toc_baselevel` is very long and potentialy mixup with baselevel 
* `toc_startlevel` is very long and potentialy mixup with baselevel 


@waylan I know you find this feature "completely unnecessary", so I am very glad that you are dealing with me;)
But I think you, as maintainer, should decide which phrase fits best in the whole project.

**Comment by waylan** [comment]
Your recent changes are a big improvement. They helped bring the new feature into focus for me. Thank you.

I agree that `toc_height` feels like the wrong name for this. Personally I would favor `toc_top`.

However, adding this feature introduces another problem, in that `toc_depth` is now misleading. A user might reasonably assume that the word "depth" refers to the total number of levels. So for example, she might expect the following settings: `toc_top=2, toc_depth=2` to result in a TOC which included levels `h2` and `h3` for a **depth of 2**. But, in fact, `toc_depth` limits the bottom to `h2` and the "depth" (or number of levels) of the TOC is no longer `2`, but `1`. Previously, when the "top" couldn't be altered, "depth" and "bottom" were always synonymous. But once the "top" is altered, "depth" and "bottom" are not the same any more. Therefore, `toc_depth` is a misleading name for the setting as it currently behaves.

I see a few ways forward:

1. Alter `toc_depth` to be the actual "depth" (number of levels) of the TOC. Perhaps do something like `bottom = top + depth - 1` (so for `toc_top=2, toc_depth=2`, `bottom` would be `3`).  For existing users who are not using `toc_top`, this changes nothing in the actual output. And as `toc_top` is new, the new behavior of `toc_depth` will be taken into consideration for any users of the new `toc_top` setting. 
2. Leave `toc_depth` behavior as-is, but deprecate the name and rename it to `toc_bottom`. This would be the most disruptive to existing users, which makes it my least favorite option. But having the two settings names "top" and "bottom" is attractive. 
3. Implement a range instead. If `toc_depth` is an integer, maintain the old behavior. If `toc_depth` is a range (`'t-b'` where `t` and `b` are integers), then treat the range as top and bottom.

Both for options 2 and 3, we would also need to handle the case where a user assigns values such that `bottom < top`. Of course, that would be an error. But in option 1 that is never a factor. As long as `depth` is any whole number between `1-6`, we will always get `bottom > top`. Of course, as HTML never drops below `h6` we don't actually care if `bottom > 6`.

**Review by waylan — APPROVED** [review]

**Comment by klml** [comment]
I like 3. best. With `toc_depth` optional as integer (as it was) or as range.

But ```t-b``` is not actually a data type, but a division. Could I use a tuple (```toc_depth = 2, 6```) instead in this case?

**Comment by waylan** [comment]
As a reminder, users don't always provide config settings in Python. Sometimes they might be using a YAML file or similar. So specifically requiring a Python `tuple` might be a stretch. However, any iterable of length 2 might work.

Personally I was thinking a string with the format `"t-b"` would make the most sense. If an integer is provided, do old behavior. If a string is provided, it must be two integers separated by a hyphen. Anything else would be an error.

File: docs/extensions/toc.md
Comment (waylan) [inline docs/extensions/toc.md]: I'm assuming you are trying to indicate that the range should be wrapped in quotes. However, triple backpacks won't give you that. You need to actually use quotes: `"2-5"`. And as a string is required, the docs should probably mention that the range needs to be a string consisting of "Two digits separated by a hyphen".

Also it is spelled "sep**a**rated", not "seperated."

File: markdown/extensions/toc.py
Comment (waylan) [inline markdown/extensions/toc.py]: Note that `self.toc_depth` could reasonably be a string which contains a single integer. Testing for an `int` here will fail in many cases. You probably need to do something like:

```python
try:
    toc_bottom = int(self.toc_depth)
    toc_top = 1
except ValueError:
    toc_top, toc_bottom = [int(t), int(b) for t, b in self.toc_depth.split('-')]
```

And that last line should perhaps be wrapped in its own try/except block to catch and raise an error if a bad range is provided.

File: markdown/extensions/toc.py
Comment (waylan) [inline markdown/extensions/toc.py]: See my previous comment about including quotes to indicate the range should be a string.

**Comment by klml** [comment]
I implemented the optional range for toc_depth. I tried to find a more elegant method than two if statements, but I hope this is ok.

**Review by waylan — CHANGES_REQUESTED** [review]
Just a few tweaks and this should be good.

**Comment by klml** [comment]
Using "try" is a good idea, sorry I missed this before. 

* I separated and duplicated the continue if clause, so we don't need to define toc_top = 1 in the integer case. 
* I moved the split from the loop over all headings to outside. 
* I defined the except for a TypeError and a ValueError. Should I use an global except to catch both?

I also thought about using the AttributeError with the integer case:

```
try:
    toc_top, toc_bottom = self.toc_depth.split('-')
    if int(el.tag[-1]) < int(toc_top) or int(el.tag[-1]) > int(toc_bottom):
        continue
except AttributeError:
    if int(el.tag[-1]) > int(self.toc_depth):
        continue
except ValueError:
    pass
```
I am not sure whats better


> And that last line should perhaps be wrapped in its own try/except block to catch and raise an error if a bad range is provided.

At the moment I have no error handling, but a wrong value just gives you the whole Table of Contents and behaves like the toc_depth is not set at all.

I think a warning should be very useful. But as long as I can see, there is no explicit error or warning-handling in Python-Markdown. When I set toc_depth to "foobar" in the existing code I get ```ValueError: invalid literal for int() with base 10: 'foobar'```. Or did I misunderstood you? 
When I have 



Thank you for your patience with this issue;)

**Comment by klml** [comment]
Using "try" is a good idea, sorry I missed this before. 

* I separated and duplicated the continue if clause, so we don't need to define toc_top = 1 in the integer case. 
* I moved the split from the loop over all headings to outside. 
* I defined the except for a TypeError and a ValueError. Should I use an global except to catch both?


I also thougt about using the AttributeError with the integer case:

```
try:
    toc_top, toc_bottom = self.toc_depth.split('-')
    if int(el.tag[-1]) < int(toc_top) or int(el.tag[-1]) > int(toc_bottom):
        continue
except AttributeError:
    if int(el.tag[-1]) > int(self.toc_depth):
        continue
except ValueError:
    pass
```
I am not sure what is better.


> And that last line should perhaps be wrapped in its own try/except block to catch and raise an error if a bad range is provided.

At the moment I have no error handling, but a wrong value just gives you the whole Table of Contents and behaves like the toc_depth is not set at all.

I think a warning should be very useful. But as long as I can see, there is no explicit error or warning-handling in Python-Markdown. When I set toc_depth to "foobar" in the existing code I get ```ValueError: invalid literal for int() with base 10: 'foobar'```. Or did I missunderstood you? 
When I resolve this, I get also these codecov.io errors.

Thank you for your patience with this issue;)

File: markdown/extensions/toc.py
Comment (waylan) [inline markdown/extensions/toc.py]: If we're going to split the range here, why not due the type conversion here as well. When we don't need all the try/except blocks later. I realize this is a different suggestion than previously, but previously I envisioned a couple lines at the top of the block of code which did the type conversion, then the actual conditionals wouldn't be so complicated.

Also wherever we do the type conversion we should define `top` and `bottom` as internal vars for later use in the conditionals. That should eliminate the need for try/except blocks in the conditionals.

**Review by waylan — CHANGES_REQUESTED** [review]

File: markdown/extensions/toc.py
Comment (klml) [reply markdown/extensions/toc.py]: I implemented your suggestions and now this looks much simpler for me.
Without the try/except blocks we lose catching "bad range errors". I am good with that.

File: markdown/extensions/toc.py
Comment (waylan) [inline markdown/extensions/toc.py]: First of all, we don't need to assign `self.toc_depth` as we don't need it anymore. `self.toc_top` and `self.toc_bottom` serve that purpose just fine. In the next few lines we can just reference `config["toc_depth"]`.

Sorry I missed this last time, but `if type(foo) is bar` is not the correct way to do a type test in Python. You should use `if isinstance(foo, bar)`, which allows users to pass in subclasses of the specified type (in other words, support duck-typing).

Thirdly, this does not account for `self.toc_depth` being a single integer within a string (`"2"`). 

There are a couple ways we could address this:

1. We could do what I suggested the first time:

    ```python
    try:
         self.toc_bottom = int(config["toc_depth"])
         self.toc_top = 1
    except ValueError:
         self.toc_top, self.toc_bottom = [int(x) for x in config["toc_depth"].split('-')]
    ```

2. Or we could check for a hyphen:

    ```python
    if isinstance(config["toc_depth"], basestring) and '-' in config["toc_depth"]:
        self.toc_top, self.toc_bottom = [int(x) for x in config["toc_depth"].split('-')]
    else:
        self.toc_top = 1
        self.toc_bottom = int(config["toc_depth"])

Notice that on both solutions, we explicitly force all values to integers. Therefore, we do not need to do that later when we use the values.

As it turns out, I like the second solution better as it keeps the old behavior and simply does a conditional check for the new syntax. Any bad data will still act exactly the same as it would have before the change. That's also why I changed it to check for a `basestring`. All string subclasses will be accepted as a range if they contain a hyphen. Whereas all other types are presumed to be an integer (the old behavior).

Comment (waylan) [reply markdown/extensions/toc.py]: I'd make these changes myself, but you used your `master` branch which makes it difficult for me to push any changes to (and I'd rather not open a new PR). If you apply the following diff exactly as-is, I'll merge this PR:

```diff
--- a/markdown/extensions/toc.py
+++ b/markdown/extensions/toc.py
@@ -135,12 +135,11 @@ class TocTreeprocessor(Treeprocessor):
         if self.use_permalinks is None:
             self.use_permalinks = config["permalink"]
         self.header_rgx = re.compile("[Hh][123456]")
-        self.toc_depth = config["toc_depth"]
-        if type(self.toc_depth) is not int:
-            self.toc_top, self.toc_bottom = self.toc_depth.split('-')
+        if isinstance(config["toc_depth"], basestring) and '-' in config["toc_depth"]:
+            self.toc_top, self.toc_bottom = [int(x) for x in config["toc_depth"].split('-')]
         else:
             self.toc_top = 1
-            self.toc_bottom = self.toc_depth
+            self.toc_bottom = int(config["toc_depth"])
 
     def iterparent(self, node):
         ''' Iterator wrapper to get allowed parent and child all at once. '''
@@ -240,7 +239,7 @@ class TocTreeprocessor(Treeprocessor):
         for el in doc.iter():
             if isinstance(el.tag, string_type) and self.header_rgx.match(el.tag):
                 self.set_level(el)
-                if int(el.tag[-1]) < int(self.toc_top) or int(el.tag[-1]) > int(self.toc_bottom):
+                if int(el.tag[-1]) < self.toc_top or int(el.tag[-1]) > self.toc_bottom:
                     continue
                 text = ''.join(el.itertext()).strip()
```

Comment (klml) [reply markdown/extensions/toc.py]: @waylan thanks for your help. I used your suggestions, but it failed on

```
if isinstance(config["toc_depth"], basestring) and '-' in config["toc_depth"]:
```
 with: _NameError: name 'basestring' is not defined_

'basestring' is python2 only, can I use [string_types from six](https://pythonhosted.org/six/#six.string_types) instead?

PS: I don't want to push non working code, so I will wait for you.

Comment (waylan) [reply markdown/extensions/toc.py]: Sorry, I forgot about that. six is why overkill for what we need. You can use `markdown.util.string_type`, which is already imported by the module as `string_type`. So just replace `basestring` with `string_type`.

Comment (klml) [reply markdown/extensions/toc.py]: @waylan Oh i missed looking into markdown/util.py, this is great;) I committed your suggestions.

**Review by waylan — APPROVED** [review]

**Comment by waylan** [comment]
This looks good. Thank you. It just needs an item added to the [3.1 release notes][1] under "New Features".

[1]: https://github.com/Python-Markdown/markdown/blob/master/docs/change_log/release-3.1.md

**Pull request merged**
